File: Week-1-April-1st-7th/Group-anagrams/group-anagrams.cpp
// Time complexity: O(k*nlogn) : k: the numbers of elements of strs and n is maximum length in a string of strs
// Space complexity: O(n)
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> m;
        for(auto s : strs){
            string tmp = s;
            sort(tmp.begin(), tmp.end()); //O(nlogn)
            m[tmp].push_back(s);
        }
        vector<vector<string>> res;
        for(auto pp : m){
            res.push_back(pp.second);
        }
        return res;
    }
};


// Time complexity: O(n*k)
//to_string converts a integer to string
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> m;
        for(auto s : strs){
            int a[26]  = {0};
            string key = "";
            for(auto c : s){
                a[c - 'a'] += 1;
            }
            for(int i = 0 ; i < 26; i++){
                key += '#' + to_string(a[i]);
            }
            m[key].push_back(s);
        }
        vector<vector<string>> ans;
        for(auto pp : m){
            ans.push_back(pp.second);
        }
        return ans;
    }
};

File: Week-1-April-1st-7th/Group-anagrams/group-anagrams.py
# Time complexity: O(k*nlogn)
"""
import collections
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        d = collections.defaultdict(list)
        for s in strs:
            tmp = sorted(s)
            d[tuple(tmp)].append(s)
        ans = []
        for v in d.values():
            ans.append(v)
        return ans
"""

# Time complexity: O(n*k) : k is length of strs and n is a maximum length of a elements of strs
import collections
class Solution:
    def groupAnagrams(self, strs):
        d = collections.defaultdict(list)
        for s in strs:
            check = [0]*26
            for c in s:
                check[ord(c) - ord('a')] += 1
            d[tuple(check)].append(s)
        
        return d.values()

strs = ['ate', 'eat', 'tae', 'uefg','geuf', 'abc']
a = Solution()
ans = a.groupAnagrams(strs)
print(ans)